### PR TITLE
Implement an online-only view layer

### PR DESCRIPTION
**Pull request by Game4all**
# Summary
This implements a `OnlineViewLayer` class geared towards use within online overlays by providing them a diming layer blocking interactions with the overlay contents and displays a centered login placeholder if the local user isn't logged in.

## How it looks with the current online overlays:
Beatmap Listing | Dashboard | Rankings
-----------------|------------|----------
listing | dashboard | rankings

### CODE REVIEW

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewLayer.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            AddAssert("content is dimmed", () => con.Colour != Color4.White);
+            AddAssert("loading animation is not visible", () => !view.LoadingSpinner.IsPresent);
+        }
+
+        [Test]
+        public void TestConnectingStateVisibility()
+        {
+            AddStep("set status to connecting", () => ((DummyAPIAccess)API).State = APIState.Connecting);
+            AddWaitStep("wait for animation to finish", 5);
+            AddAssert("content is dimmed", () => con.Colour != Color4.White);
```
Comment (bdach) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewLayer.cs]: Might also be worth checking that the button underneath is not interactable.

File: osu.Game/Online/OnlineViewLayer.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        public override bool HandleNonPositionalInput => false;
+
+        protected override bool Handle(UIEvent e) => true;
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Children = new Drawable[]
+            {
+                placeholder = new LoginPlaceholder(placeholderMessage)
```
Comment (bdach) [inline osu.Game/Online/OnlineViewLayer.cs]: Doesn't really make sense to have a `LoginPlaceholder` be used here, in a general `OnlineViewLayer` - should be either abstracted away to just a `Placeholder` or renamed to reflect the new more-general purpose.

File: osu.Game.Tests/Visual/Online/TestSceneOnlineViewLayer.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    view = new TestOnlineViewLayer("Please sign in to view content!", con)
+                }
+            };
+        }
+
+        [Test]
+        public void TestOfflineStateVisibility()
+        {
+            AddStep("set status to offline", () => ((DummyAPIAccess)API).State = APIState.Offline);
+            AddWaitStep("wait for animation to finish", 5);
```
Comment (bdach) [inline osu.Game.Tests/Visual/Online/TestSceneOnlineViewLayer.cs]: Probably more reliable to use an `UntilStep`. No need to expose the placeholder either, can use `ChildrenOfType<>`.

File: osu.Game/Online/OnlineViewLayer.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A layer which displays on top of content which require a local user to be logged in.
+    /// Blocks input to the underlying content, dims the optional target content and displays a login placeholder when the user isn't logged in.
+    /// </summary>
+    public class OnlineViewLayer : VisibilityContainer, IOnlineComponent
+    {
+        protected LoadingSpinner LoadingSpinner;
```
Comment (bdach) [inline osu.Game/Online/OnlineViewLayer.cs]: If this is exposed just for the test, might as well use `ChildrenOfType<>` for that too instead and not expose.

**Review by bdach — CHANGES_REQUESTED** [review]
Out of curiosity, have you tried to apply this in more places and is changing to a layer a conclusion you've arrived to after trying that? I'm a bit worried that we'll flip-flop between the two variants.

**Comment by Game4all** [comment]
It may be better as you pointed out to let both exist as variants.

**Comment by bdach** [comment]
If that ends up being the case they should share logic as much as possible, but then I don't think they *both* should be permitted to exist if it's at all avoidable.

**Comment by Game4all** [comment]
The only place where I found an usage for OnlineViewLayer for now is for online overlays where it is preferable to dim content instead of hiding it. 
The logic could have been embedded in shared a class if all online-overlays inherited a same class but since the ChatOverlay doesn't, it needs to exist as a variant to OVC.

**Comment by peppy** [comment]
@Game4all i know this is a bit out of date now and didn't really get reviewed in time. We now have an `OnlineView<>` class for the new dashboard setup which can potentially become one of these. Did you want to give it a shot and see if it fits well?

**Comment by Game4all** [comment]
Sounds fine, will give it a shot to try integrating offline behaviour handling. It get the feeling that this may basically supersede what this PR proposed if it turns to work well. (Although it may require quite a bit of refactoring to make use of it in all online-only overlays)

**Comment by Game4all** [comment]
@peppy After messing a bit with `OverlayView<>`, I get the feeling that `OverlayView<>` can definitely supersede this PR contents if it were to have the offline logic handling added directly to it, altough it may require tweaking transitions. Thus, I'm closing this and will start from scratch implementing the behaviour proposed by this PR in `OverlayView<>`.

**Pull request closed without merging**